inference-engine/include/details/ie_mmap_allocator.hpp

```diff
@@ -0,0 +1,276 @@
+// Copyright (C) 2018-2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief The header file defines utility PreAllocator class
+ * 
+ * @file ie_pre_allocator.hpp
+ */
+#pragma once
+
+#include <details/ie_exception.hpp>
+#include <memory>
+
+#include "ie_allocator.hpp"
+
+#ifdef _WIN32
+#include <windows.h>
+#define MAP_FAILED nullptr
+#else
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#define INVALID_HANDLE_VALUE -1
+#endif
+
+namespace InferenceEngine {
+namespace details {
+/*
+ * @brief This is a helper class to wrap memory mapped files
+ */
+class MmapAllocator : public IAllocator {
+public:
+    using Ptr = std::shared_ptr<MmapAllocator>;
+
+#if defined(ENABLE_UNICODE_PATH_SUPPORT) && defined(_WIN32)
+    using path_type = std::wstring;
+#else
+    using path_type = std::string;
+#endif
+
+    MmapAllocator(const path_type& path, size_t offset = 0, LockOp lock = LOCK_FOR_READ)
+        : _path(path), _offset(offset), _size(0), _lock(lock), _data(nullptr) {
+    }
+
+    /**
+     * @brief Locks a handle to heap memory accessible by any memory manipulation routines
+     * @return The generic pointer to a memory buffer
+     */
+    void* lock(void* handle, LockOp) noexcept override {
+        return handle;
+    }
+    /**
+     * @brief The MmapAllocator class does not utilize this function
+     */
+    void unlock(void*) noexcept override {}  // NOLINT
+
+    /**
+     * @brief Returns a pointer to mapped memory
+     * @param size Size in bytes
+     * @return A handle to the preallocated memory or nullptr
+     */
+    void* alloc(size_t size) noexcept override {
+        map(_path.c_str(), _offset, size, _lock);
```

Comment:
The map() function can throw exception but this current function defined as noexcept. So we'll crash if exception is happened.